pkg/istio-agent/README.md

```diff
@@ -0,0 +1,82 @@
+# Istio Agent
+
+This document describes the internal architecture of Istio agent.
+
+## High Level overview
+
+![High Level overview](docs/overview.svg)
+
+At a high level, the Istio agent acts as an intermediate proxy between Istiod and Envoy. This is done
+at two levels. For distributing workload certificates, Envoy will send [SDS](https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret)
+requests to the agent, causing the agent to submit a CSR to the configured CA (generally Istiod). For other configuration,
+Envoy will send [ADS](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/dynamic_configuration#aggregated-xds-ads)
+requests to the agent, which will be forwarded to the configured discovery server (general Istiod).
+
+## CA Flow
+
+![CA Flow](docs/ca.svg)
+
+A single SDS request from Envoy goes through a few different layers in istio-agent.
+
+1. First, the request is handled by the SDS server. This is mostly just an intermediate translation layer exposing
+   the `SecretManager` to Envoy, without much business logic. For each resource requested by Envoy, the SDS server
+   will call `SecretManager.GenerateSecret(resourceName)`
+1. When `GenerateSecret` is called, the `SecretManager` is expected to return a new certificate. This can occur in a few ways.
+   1. The most common method (pictured above) is to sign a new certificate by calling the configured. Typically, this is Istiod.
+   1. If the certificate not yet expired, the `SecretManager` also can return a cache response. In practice, this would
+       only happen in cases where Envoy were to re-request a resource, which is fairly rare.
```

Comment:
Envoy hot restart is no longer used. Its still possible though, if a user disables mTLS then re-enables it I think we would end up with no SDS connections open while its disabled

Reply:
Not clear if it's worth caching and returning the old cert, the system should be able to handle restarts of envoys when pods are rolled out.

Reply:
I originally had the caching removed. The only reason I added it back is so to avoid 2x CSRs on startup (the "warming" one we send immediately, then in response to the SDS request). We also already have to cache the root cert (unless we send a full CSR on a request for it I suppose, which would be a bit odd/slow), so it wasn't too bad.

So for those 2 reasons I decided to add it back. I think the small complexity is worth reduced CSRs, unless you have a suggestion to have no cache and no repetitive CSRs?